### ISSUE TITLE
JsonPatchDocument should use System.Text.Json

### ISSUE THREAD

**Issue by xperiandri** [dotnet/aspnetcore#24333]
Related to https://github.com/dotnet/aspnetcore/issues/16968
Will you accept pull-request with this feature?

**Comment by xperiandri** [comment]
I propose to:
* add `IOperation` interface
* create new package `Microsoft.AspNetCore.JsonPatch.Newtonsoft` referencing https://github.com/dotnet/aspnetcore/blob/master/src/Features/JsonPatch/src/Operations/Operation.cs as link with conditional compilation namespace
* add package `Microsoft.AspNetCore.JsonPatch.System.Text` referencing https://github.com/dotnet/aspnetcore/blob/master/src/Features/JsonPatch/src/Operations/Operation.cs as link with conditional compilation namespace

**Comment by mkArtakMSFT** [comment]
Thanks for your willingness to help.
We would prefer to think more about this first and decide on the direction we want to go before asking for help.

**Comment by null-d3v** [comment]
There have been many issues reported regarding json patch still being tied to Newtonsoft.Json deserialization. The issues are caused by adverse effects of the association, or directly indicating dissatisfaction with it:
https://github.com/dotnet/aspnetcore/issues/12994
https://github.com/dotnet/aspnetcore/issues/13938
https://github.com/dotnet/aspnetcore/issues/14035
https://github.com/dotnet/aspnetcore/issues/16968
https://github.com/dotnet/aspnetcore/issues/22216
https://github.com/dotnet/aspnetcore/issues/22620

System.Text.Json has merits over Newtonsoft.Json [purported by Microsoft](https://devblogs.microsoft.com/dotnet/try-the-new-system-text-json-apis/#user-content-the-future-of-json-in-net-core-30) with indication to fully remove the Json.NET dependency from ASP.NET Core.
> Thus, we want to remove the Json.NET dependency from ASP.NET Core 3.0, so that customers can choose which version to use, without fearing they might accidentally break the underlying platform.

Configuration of json patch to only use Newtonsoft.Json processing for patch requests is contrived, requiring building an entire separate service collection with complete MVC to pull an input formatter with appropriate DI.

To indicate the intent of complete System.Text.Json support in ASP.NET Core, not have the support, and then to consistently delay in the face of consistent issues is extremely discouraging:
https://github.com/dotnet/aspnetcore/issues/12994#issuecomment-521031088
https://github.com/dotnet/aspnetcore/issues/14035#issuecomment-532773409
https://github.com/dotnet/aspnetcore/issues/14035#issuecomment-552260138
https://github.com/dotnet/aspnetcore/issues/16968#issuecomment-620212910

**Comment by ylabade** [comment]
Do you have any news regarding this issue? Do you accept help to implement this feature?

**Comment by erwan-joly** [comment]
I wonder if the easiest way to have that feature added wouldn't be to fork the feature project https://github.com/dotnet/aspnetcore/commits/master/src/Features/JsonPatch in another repo
make it works without newtonsoft 
then ask dotnet/aspnetcore if they want to replace newtonsoft json patch to that one

**Comment by apathytomorrow** [comment]
This is bit of an annoying oversight. Any news?

**Comment by erwan-joly** [comment]
https://github.com/gregsdennis/json-everything

**Comment by natelaff** [comment]
+1

**Comment by sulmar** [comment]
Do you have any plans to use JsonPatchDocument with System.Text.Json in upcoming .NET 6 ?
It will be very handy in Minimal API.

**Comment by Havunen** [comment]
Has anybody found good open source json patch library for system text json?

**Comment by erwan-joly** [comment]
> Has anybody found good open source json patch library for system text json?

https://github.com/dotnet/aspnetcore/issues/24333#issuecomment-747723722

**Comment by winromulus** [comment]
A lot of projects have dropped support for Newtonsoft.Json (especially for annotations)  in favor of System.Text.Json
Some projects, like the Kubernetes C# client rely heavily on support for JSON Patch.
Is it possible to re-evaluate the priority on this in light of increased requests for support on this topic? It's been years since it was considered low-priority since few projects had been ported to System.Text.Json at the time.

**Comment by Grizzlly** [comment]
@pranavkm I think this is worth looking into. It is a bummer that we have to rely on Newtonsoft.

The [code in the documentation](https://docs.microsoft.com/en-us/aspnet/core/web-api/jsonpatch?view=aspnetcore-6.0#json-patch-addnewtonsoftjson-and-systemtextjson) throws `ASP0000` warning. This should be mentioned in the docs imo.

**Comment by beachwalker** [comment]
Still a issue? Keeps us from switching.

**Comment by Grizzlly** [comment]
> Still a issue? Keeps us from switching.

Yes.

**Comment by rqueizan** [comment]
Install this package Microsoft.AspNetCore.Mvc.Formatters.Json

**Comment by erwan-joly** [comment]
> Install this package Microsoft.AspNetCore.Mvc.Formatters.Json for System.Text.Json

This seems to use newtonsoft from Microsoft.AspNetCore.JsonPatch

**Comment by rqueizan** [comment]
> > Install this package Microsoft.AspNetCore.Mvc.Formatters.Json for System.Text.Json
> 
> This seems to use newtonsoft from Microsoft.AspNetCore.JsonPatch

![image](https://user-images.githubusercontent.com/18214796/157108270-b1efd03e-dcc3-4add-9d90-5a6fb7df4e95.png)

![image](https://user-images.githubusercontent.com/18214796/157108371-2f4a5cdb-3b66-4bbc-93a4-6e6c4d6ca78e.png)

![image](https://user-images.githubusercontent.com/18214796/157108498-6c740112-d67d-4707-a006-b8e04afbacf5.png)

I am not using Newtonsoft.Json!

**Comment by erwan-joly** [comment]
> > > Install this package Microsoft.AspNetCore.Mvc.Formatters.Json for System.Text.Json
> > 
> > 
> > This seems to use newtonsoft from Microsoft.AspNetCore.JsonPatch
> 
> ![image](https://user-images.githubusercontent.com/18214796/157108270-b1efd03e-dcc3-4add-9d90-5a6fb7df4e95.png)
> 
> ![image](https://user-images.githubusercontent.com/18214796/157108371-2f4a5cdb-3b66-4bbc-93a4-6e6c4d6ca78e.png)
> 
> ![image](https://user-images.githubusercontent.com/18214796/157108498-6c740112-d67d-4707-a006-b8e04afbacf5.png)
> 
> I am not using Newtonsoft.Json!

Indirectly your are https://www.nuget.org/packages/Microsoft.AspNetCore.JsonPatch/

**Comment by rqueizan** [comment]
> > > > Install this package Microsoft.AspNetCore.Mvc.Formatters.Json for System.Text.Json
> > > 
> > > 
> > > This seems to use newtonsoft from Microsoft.AspNetCore.JsonPatch
> > 
> > 
> > ![image](https://user-images.githubusercontent.com/18214796/157108270-b1efd03e-dcc3-4add-9d90-5a6fb7df4e95.png)
> > ![image](https://user-images.githubusercontent.com/18214796/157108371-2f4a5cdb-3b66-4bbc-93a4-6e6c4d6ca78e.png)
> > ![image](https://user-images.githubusercontent.com/18214796/157108498-6c740112-d67d-4707-a006-b8e04afbacf5.png)
> > I am not using Newtonsoft.Json!
> 
> Indirectly your are https://www.nuget.org/packages/Microsoft.AspNetCore.JsonPatch/

You are rigth
      "Microsoft.AspNetCore.JsonPatch/2.2.0": {
        "dependencies": {
          "Microsoft.CSharp": "4.7.0",
          "Newtonsoft.Json": "11.0.2"
        },
        "runtime": {
          "lib/netstandard2.0/Microsoft.AspNetCore.JsonPatch.dll": {
            "assemblyVersion": "2.2.0.0",
            "fileVersion": "2.2.0.18316"
          }
        }
      },

**Comment by OpenSpacesAndPlaces** [comment]
Ran into this too. 

Having to manually read the body and parse with Newtonsoft since System.Text.Json doesn't work as the controller default.

**Comment by k-seanon** [comment]
also having to steer away from system.text.json because of this

**Comment by pdevito3** [comment]
+1 

Surprised this is marked as `affected-few` and wasn't even considered for .NET 7. [The solution in the MS docs](https://docs.microsoft.com/en-us/aspnet/core/web-api/jsonpatch?view=aspnetcore-6.0#add-support-for-json-patch-when-using-systemtextjson) doesn't even work for me.

@mkArtakMSFT I know you were thinking through solutions, but do you guys have any more recent updates on this?

**Comment by pdevito3** [comment]
So I was just playing around with what it might look like to convert my controllers to minimal APIs and noticed that we don't currently have `MapPatch`, but that it's coming in .NET 7.

@davidfowl @captainsafia I saw that you guys had some traffic on that in #36198 and want to raise the flag on this issue in case you're not aware. Thinking many people will reach for a `Patchdoc` when setting it up and this will block that option for them as far as I can tell -- unless there's a way to set that patch endpoint to use newtonsoft explicitly but even so, that would be somewhat hacky.

**Comment by davidfowl** [comment]
@pdevito3 Thanks! I'm not sure we'll be able to pull this off for .NET 7 as patch support for System.Text.Json is currently unplanned but this is something we should get ahead of at least in the docs. Maybe we can write a sample converter to make this work in the mean time.

**Comment by Havunen** [comment]
I started implementing this myself, its based on aspnet version. https://github.com/Havunen/SystemTextJsonPatch

**Comment by davidfowl** [comment]
@Havunen WONDERFUL! We should point people to this for now. Are there any other implementations that you are aware of?

**Comment by erwan-joly** [comment]
> @Havunen WONDERFUL! We should point people to this for now. Are there any other implementations that you are aware of?

https://github.com/dotnet/aspnetcore/issues/24333#issuecomment-747723722

**Comment by natilivni** [comment]
Any update on this?

**Comment by davidfowl** [comment]
We are not doing this for .NET 7, the work isn't planned and doesn't fit into the release based on the other work we're already doing.

**Comment by stevendarby** [comment]
Could MS please clarify what their commitment to AspNetCore.JsonPatch is, as it's looking pretty shaky to me. Reasons for thinking this:
- STJ was released with .NET Core 3.0 in late 2019 and with no support now or in upcoming 7.0, late 2023 is the earliest it could be supported. 5 years!
- Bugs and requests against it just seem to get backlogged (including two of mine) despite very little activity in that area. I think JsonPatch sharing a backlog with core ASP.NET Core means it'll always be deprioritised unless a conscious push is made.
- Key MS developers are endorsing 3rd party libraries instead 😟

**Comment by Havunen** [comment]
> Bugs and requests against it just seem to get backlogged (including two of mine) despite very little activity in that area. I think JsonPatch sharing a backlog with core ASP.NET Core means it'll always be deprioritised unless a conscious push is made 

Can you try out SystemTextJsonPatch? I believe I have fixed both the issues you reported about AspNetCore.JsonPatch.
I know its not the official Microsoft package but personally I just needed a package that works and does not depend on Newtonsoft.

**Comment by hexjourney** [comment]
This issue don't impact a small number of customers...

**Comment by carlblanchard** [comment]
Why isn't System.Text.Json adopting JsonPatchDocument? 
It's a web standard technology, after all, if you want people to adopt System.Text.Json over NewtonSoft you're going to address people's demands not just ignore them and say it's on the backlog.

**Comment by lennieh** [comment]
Minimal API in .Net 7 supports patch operations. Has anyone worked out how to use this with JsonPatchDocument??

**Comment by sulmar** [comment]
You can do it, step by step, this way, for example:

``` bash
dotnet add package Newtonsoft.Json
``` 

``` csharp
app.MapPatch("/", async (HttpContext context) =>
{
    if (context.Request.HasJsonContentType())
    {       
        using var reader = new StreamReader(context.Request.Body);
        var json = await reader.ReadToEndAsync();
        var doc = JsonConvert.DeserializeObject<JsonPatchDocument>(json);

        var originalDocument = new Models.Document();  // change to get an entity from database
        doc.ApplyTo(originalDocument);

        return Results.Ok(originalDocument);        
    }
    return Results.BadRequest();
});


public class Document
{
    public string Baz { get; set; } = "qux";
    public string[] Hello { get; set; }
    public string Foo { get; set; } = "bar";
}

```

With this solution Newtonsoft.Json is used only for single endpoint.

Maybe someone knows a better solution?

**Comment by payoff** [comment]
> Maybe someone knows a better solution?
Probably use: 
https://github.com/Havunen/SystemTextJsonPatch

**Comment by sulmar** [comment]
Thank to @deleteLater we can use in Minimal Api:

``` csharp
app.MapPatch("/", async ([FromBody] JsonElement jsonElement) =>
{
      var json = jsonElement.GetRawText();

      var doc = JsonConvert.DeserializeObject<JsonPatchDocument>(json);

      var originalDocument = new JsonPatchDocumentMinimalApi.Models.Document();
      doc.ApplyTo(originalDocument);

      return Results.Ok(originalDocument);
});
```

**Comment by tjmcdonough** [comment]
@Havunen solution works very well. I have just given the GitHub project a star :) Highly recommend we all use this until Microsoft comes up with a solution

**Comment by zinov** [comment]
> I'm looking for better implementation yet.
> 
> I thought about using my own implemention of IModelBinder but in the minimal api seems not to be supported.
> 
> In .NET 7 for Custom Binding we can use only TryParse or BindAsync methods: https://learn.microsoft.com/en-us/aspnet/core/fundamentals/minimal-apis/parameter-binding?view=aspnetcore-7.0#custom-binding
> 
> It is easy to add for own model but not for JsonPatchDocument.

this is something that should be coming with the framework instead of doing a custom model binder to map all parameters. Microsoft should follow the standard: https://datatracker.ietf.org/doc/html/rfc6902

**Comment by Arkovski** [comment]
[mkArtakMSFT](https://github.com/mkArtakMSFT) you're thinking quite lot, maybe time to do anything? :)

**Comment by rohitvipin** [comment]
Happy 4 year anniversary for a very important issue :(

**Comment by DanielThyselius** [comment]
> > Bugs and requests against it just seem to get backlogged (including two of mine) despite very little activity in that area. I think JsonPatch sharing a backlog with core ASP.NET Core means it'll always be deprioritised unless a conscious push is made
> 
> Can you try out SystemTextJsonPatch? I believe I have fixed both the issues you reported about AspNetCore.JsonPatch. I know its not the official Microsoft package but personally I just needed a package that works and does not depend on Newtonsoft.

I've tried this out and it worked out of the box, brilliant work!

**Comment by lincoln42** [comment]
Its niggling and neglected issues like this that cause many enterprise projects to abandon using the .Net framework entirely and move other API frameworks like Python FastAPI (https://medium.com/@david.danier/how-to-handle-patch-requests-with-fastapi-c9a47ac51f04) or Java SpringBoot 3 (https://www.baeldung.com/spring-rest-json-patch). 

.Net is regarded as incomplete for enterprise work due to such issues and makes advocating for it's continued use in the enterprise really difficult when competing frameworks offer a friction free developer experience.

**Comment by stevendarby** [comment]
> STJ was released with .NET Core 3.0 in late 2019 and with no support now or in upcoming 7.0, late 2023 is the earliest it could be supported. 4 years!

2024 update: it's not coming in 9.0, so the earliest possible support is now 10.0 in late 2025, 6 years after STJ.

**Comment by mikekistler** [comment]
This issue has been open for a long time and has many comments and upvotes, so we're evaluating this as a potential feature for .NET 10.

There are a number of third party packages, several mentioned in this issue, that provide support for JSON Patch with System.Text.Json. In light of that, we'd like to hear from users whether it's necessary to add this support into the core framework or if support with a third party package is acceptable.

Bear in mind that adding this support to the core framework will require resources that could otherwise be applied to other features. So think of the most important feature you're hoping for in .NET 10, and if built-in JSON Patch support is more important than that, we definitely want to hear that perspective. And we also want to hear the contrary -- that support from a third party package is perfectly fine for you.

Please share your views on this question below, and especially the "why" behind your views, so we can apply our resources in the most impactful way for .NET users.

Thank you!

**Comment by carlblanchard** [comment]
Hi Mike,

I strongly believe that the PATCH verb and an Operation object are long overdue to be included in .NET. 
While I understand that PATCH was introduced later than other HTTP verbs, it is still an RFC standard, and its official implementation in .NET is long overdue. Standardising it would greatly improve consistency and best practices across the ecosystem.

https://datatracker.ietf.org/doc/html/rfc6902#section-4.1

**Comment by stevendarby** [comment]
Aspire.Hosting.AppHost
![image](https://github.com/user-attachments/assets/a1aa1b4a-8865-4e83-ab72-0165e6133259)

Microsoft.AspNetCore.JsonPatch
![image](https://github.com/user-attachments/assets/2ddfaa5d-697d-44cc-bbde-7a70c379c063)

**Comment by lincoln42** [comment]
I agree with Carl, since JSON patch is a standard, it should be something that has native support in the framework.

There are some good projects that offer this functionality that can benefit from Microsoft sponsorship (Havunen/SystemTextJsonPatch: SystemTextJsonPatch is a JSON Patch (JsonPatchDocument) RFC 6902 implementation for .NET using System.Text.Json<https://github.com/Havunen/SystemTextJsonPatch>)

Additionally native integration for the below in AspNetCore will be much appreciated and contribute to increased adoption of AspNetCore for microservices in the enterprise.

The below are really required to enable parity with frameworks like Spring Boot, without which .Net will continue to steadily lose ground in the enterprise where these CNCF projects are very popular and in high demand.

Apache Kafka (a good 3rd party project that would benefit from Microsoft sponsorship ch-robinson/dotnet-avro: An Avro implementation for .NET<https://github.com/ch-robinson/dotnet-avro>),

Databricks & Apache Spark (without having to pay for the CData 3rd party driver) (dotnet/spark: .NET for Apache® Spark™ makes Apache Spark™ easily accessible to .NET developers.<https://github.com/dotnet/spark>_ another project that would benefit from Microsoft sponsorship)

Apache Flink
________________________________

**Comment by OpenSpacesAndPlaces** [comment]
I think the frustration the community has with feature requests like this is that there is seemingly more mental effort spent punting vs. the time it would take to add the missing element in. To be fair though, this is hardly a problem unique to .Net Core. If you've been in software development long enough you'd had at least one "why we shouldn't bother meeting" that runs longer and uses more resources then implementation time to build something.

I get the counter argument that if there is no vetting, that leads to system bloat and what I'll call the "rural road problem" where the is more to maintain than resources can afford.

TLDR - Low effort, low maintenance, high reward features- there has to be a more efficient path to vet these and get them in.

**Comment by zinov** [comment]
This should be natively supported by Microsoft in my opinion, not by a
third party library. Adoption is going to be faster coming from Microsoft
than other libraries.

**Comment by lincoln42** [comment]
In terms of prioritising for .Net 10, native support for CNCF projects with high enterprise adoption rates like Kafka, Spark, Flink & Databricks will go a long way to increasing the adoption of .Net in the enterprise just like the move to being able to run cross platform did for the adoption of .Net Core for microservies on Kubernetes.

________________________________

**Comment by ChrisSimmons** [comment]
> This issue has been open for a long time and has many comments and upvotes, so we're evaluating this as a potential feature for .NET 10.
> 
> There are a number of third party packages, several mentioned in this issue, that provide support for JSON Patch with System.Text.Json. In light of that, we'd like to hear from users whether it's necessary to add this support into the core framework or if support with a third party package is acceptable.
> 
> Bear in mind that adding this support to the core framework will require resources that could otherwise be applied to other features. So think of the most important feature you're hoping for in .NET 10, and if built-in JSON Patch support is more important than that, we definitely want to hear that perspective. And we also want to hear the contrary -- that support from a third party package is perfectly fine for you.
> 
> Please share your views on this question below, and especially the "why" behind your views, so we can apply our resources in the most impactful way for .NET users.
> 
> Thank you!

@mikekistler It's great to see that Microsoft is looking at this.  I would prefer to see first-class support from Microsoft in the core framework and I'll try to break down my "why" reasoning as simply as possible that meet what you're looking for to justify dedicating resources.

## JSON Patch support is more important than [insert other possible new .NET features]

Why do I think the functionality itself is important?  JSON Patch support is highly important to me, and specifically more important than other possible .NET features, because I work in REST web services all-day-every-day.  This means an implementation of `PATCH` that meets an RFC standard is critical.  This isn't a corner case or nice-to-have feature.  `PATCH` is foundational and we've had an RFC available to describe this for going on 12 years now.

## Stability / Longevity of a solution

Why do I think Microsoft should provide it in the core framework as opposed to relying on third-party solutions?  Simply put here, at least from my point of view, Microsoft has done a fantastic job of keeping stability in their APIs for long periods of time.  I'm not saying you guys are perfect but overall you're d*** good.

But I'm not saying that to flatter.  A lot of my feelings on this point have to do with recent failings in trust for widely-used packages.  I'm thinking specifically of Moq and FluentAssertions.  My examples there are nice-to-haves ultimately yet still caused great consternation when they changed their license or inserted some sketchy functionality.  Something _this_ critical, IMHO, needs to be provided by a funded organization that is highly _unlikely_ to rug-pull like this.

Hope that helps your analysis of this issue.

**Comment by abratv** [comment]
i think the team need better product manager
Http patch should be basic stuff included in the core framework
The priority went to Aspire (nice to have) and "fancy" stuff with Blazor
do we really need to wait for .net 10 for this?

**Comment by mikekistler** [comment]
Thanks to all who shared your thoughts on this issue. We posted the [roadmap for .NET 10](http://aka.ms/dotnet/roadmap) earlier this week and we have included this in the proposed set of features for ASP.NET. As always, the roadmap is "an aspirational list of what we hope to get to", but given the level of frustration with this issue we are very hopeful that we can deliver this feature in .NET 10.

**Comment by lincoln42** [comment]
Thanks much appreciated. Looking forward to .Net 10.

Sent from Outlook for Android<https://aka.ms/AAb9ysg>
________________________________

**Comment by carlblanchard** [comment]
@mikekistler this is great :) Thank you.

**Comment by zinov** [comment]
Finally. This is something that is impacting every integration that is
relying on this method using rest standards. I think the community is going
to be happy with this delivery

**Comment by mkArtakMSFT** [comment]
A new System.Text.Json serialization-based implementation was now merged. Basically, there will be a new package - Microsoft.AspNetCore.JsonPatch.SystemTextJson that will ship in .NET 10 Preview 4. I will appreciate if people here could try that out once it ships. If you find things that don't work, please file new issues so that we can see and address them as necessary. Thanks again for your patience.

**Comment by wertzui** [comment]
Why is `Newtonsoft.Json` still listed as dependency in RC1?
According to this Issue, I would expect it to be gone.
https://www.nuget.org/packages/Microsoft.AspNetCore.JsonPatch/10.0.0-rc.1.25451.107#dependencies-body-tab

**Comment by stevendarby** [comment]
@wertzui see the post above yours

**Comment by wertzui** [comment]
Sorry, my bad 😄 
For anyone looking for the correct package, here is the link:
https://www.nuget.org/packages/Microsoft.AspNetCore.JsonPatch.SystemTextJson#dependencies-body-tab
